Gecko EMF Exporters (CSV ZIP, CSV flat, XLSX, ODS) - new & updated

ideas-into-software opened the pull request:
### Gecko EMF XLSX Exporter

 - XLSX Exporter for EMF with all features implemented (including links,
    freeze header row, column width adjustment, row heigh adjustment, etc.)
    
 - Various fixes & improvements

Commit:
Gecko EMF CSV Exporter, and more

 - Support for exporting EMF resources and lists of EMF objects to CSV
format in ZIP mode

 - Refactor of format-agnostic parts / common to all formats

 - Refactor of format-specific parts

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

Commit:
ODS (OpenDocument Spreadsheet) Exporter for EMF (c.d.)

 - Potential fix for performance issue - over 7 times (700+%) speed-up

 - Refactoring

 - (temporarily) large-data set (trees) for performance measurement

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

Commit:
XLSX Exporter for EMF (c.d.), and more

 - XLSX Exporter for EMF with all features implemented (including links,
freeze header row, column width adjustment, row heigh adjustment, etc.)

 - Various fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
Attached are XLSX documents generated off 3 different models.

[testExportBasicPackageEObjectsToXlsx15607552430106964369.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12571323/testExportBasicPackageEObjectsToXlsx15607552430106964369.xlsx)
[testExportUtilitiesPackageEObjectsToXlsx4936331515545625973.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12571324/testExportUtilitiesPackageEObjectsToXlsx4936331515545625973.xlsx)
[treesPackageExportEObjectsToTest14185856345420330110.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12571325/treesPackageExportEObjectsToTest14185856345420330110.xlsx)

Commit:
Gecko EMF CSV Exporter (flat mode), and more

 - Support for exporting EMF resources and lists of EMF objects to CSV
format in FLAT mode

 - Refactor of format-agnostic parts / common to all formats (c.d.),
including column headers, value cells, etc.

 - Additional validations (class hierarchy, export options'
combinations, number of columns, etc.)

 - Performance improvements for processing of large data sets (lookup
indexes, etc.)

 - Various other fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
### Gecko EMF CSV Exporter (flat mode)

 - Support for exporting EMF resources and lists of EMF objects to CSV format in FLAT mode

 - Refactor of format-agnostic parts / common to all formats (c.d.), including column headers, value cells, etc.
 
 - Additional validations (class hierarchy, export options' combinations, number of columns, etc.)
 
 - Performance improvements for processing of large data sets (lookup indexes, etc.)
 
 - Various other fixes & improvements

Commit:
Gecko EMF CSV Exporter (flat mode), and more (c.d.)

 - Bug fix for #skipFeature method

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
Attached are CSV documents generated off 3 different models - in both ZIP and flat modes.

[testExportTreesModelEObjectsToCsvFlatMode17971557483261693482.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12707224/testExportTreesModelEObjectsToCsvFlatMode17971557483261693482.csv)
[testExportUtilModelEObjectsToCsvFlatMode4058704291043845551.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12707226/testExportUtilModelEObjectsToCsvFlatMode4058704291043845551.csv)
[testExportExampleModelBasicEObjectsToCsvFlatMode13692674266014937147.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12707227/testExportExampleModelBasicEObjectsToCsvFlatMode13692674266014937147.csv)


[testExportExampleModelBasicEObjectsToCsvZipMode981403300270773804.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12707217/testExportExampleModelBasicEObjectsToCsvZipMode981403300270773804.zip)
[testExportTreesModelEObjectsToCsvZipMode14658937735468231644.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12707218/testExportTreesModelEObjectsToCsvZipMode14658937735468231644.zip)
[testExportUtilModelEObjectsToCsvZipMode9689777306655229293.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12707223/testExportUtilModelEObjectsToCsvZipMode9689777306655229293.zip)

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS)  (c.d.)

 - Additional features, including 'show URIs'

 - Various fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
@juergen-albert

With most recent commit, work on Gecko EMF Exporters (CSV, XLSX, ODS) is now complete – pending your review of output produced (see below). 

All 3 exporters are in this PR, due to large refactoring which took place, affecting all existing and new (CSV) exporters.

All important combinations of export options were tested across 3 different models (`org.gecko.emf.osgi.example.model.basic`, `org.gecko.emf.util.model`, `org.gecko.emf.trees.model`), via a total of 30+ test cases in all 3 exporters (CSV, XLSX, ODS).

To reconcile your request regarding URIs, additional export option ‘show URIs’ was added, turned on by default. So, in all columns where previously reference IDs were present, you can see URIs now.

Attached please find output from all 3 exporters: 
    • CSV (flat mode) documents generated off 3 different models
    • CSV (ZIP mode) documents generated off 3 different models
    • XLSX documents generated off 3 different models
    • ODS documents generated off 3 different models

---
### CSV (flat mode) documents generated off 3 different models

[testExportExampleModelBasicEObjectsToCsvFlatModeNonContainmenEnabled17270292410785697696.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775268/testExportExampleModelBasicEObjectsToCsvFlatModeNonContainmenEnabled17270292410785697696.csv)
[testExportExampleModelBasicResourceToCsvFlatModeDifferentRootObjectsNonContainmentDisabled3558216633382306798.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775269/testExportExampleModelBasicResourceToCsvFlatModeDifferentRootObjectsNonContainmentDisabled3558216633382306798.csv)
[testExportExampleModelBasicResourceToCsvFlatModeNonContainmentDisabled6392309440189744748.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775270/testExportExampleModelBasicResourceToCsvFlatModeNonContainmentDisabled6392309440189744748.csv)
[testExportTreesModelEObjectsToCsvFlatModeNonContainmentDisabled11125104523448083478.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775271/testExportTreesModelEObjectsToCsvFlatModeNonContainmentDisabled11125104523448083478.csv)
[testExportUtilModelEObjectsToCsvFlatModeNonContainmenEnabled14492919839323587345.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775272/testExportUtilModelEObjectsToCsvFlatModeNonContainmenEnabled14492919839323587345.csv)
[testExportUtilModelResourceToCsvFlatModeNonContainmentDisabled13799934680041756833.csv](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775273/testExportUtilModelResourceToCsvFlatModeNonContainmentDisabled13799934680041756833.csv)


---
### CSV (ZIP mode) documents generated off 3 different models

[testExportExampleModelBasicEObjectsToCsvZipModeNonContainmentDisabled9716503688394904352.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775275/testExportExampleModelBasicEObjectsToCsvZipModeNonContainmentDisabled9716503688394904352.zip)
[testExportExampleModelBasicResourceToCsvZipModeNonContainmenEnabled2544465538067109502.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775276/testExportExampleModelBasicResourceToCsvZipModeNonContainmenEnabled2544465538067109502.zip)
[testExportTreesModelEObjectsToCsvZipModeNonContainmenEnabled7577974545915050310.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775277/testExportTreesModelEObjectsToCsvZipModeNonContainmenEnabled7577974545915050310.zip)
[testExportUtilModelEObjectsToCsvZipModeNonContainmenEnabled8250948091862984091.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775278/testExportUtilModelEObjectsToCsvZipModeNonContainmenEnabled8250948091862984091.zip)
[testExportUtilModelEObjectsToCsvZipModeNonContainmentDisabled5209827492020129073.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775279/testExportUtilModelEObjectsToCsvZipModeNonContainmentDisabled5209827492020129073.zip)

---
### XLSX documents generated off 3 different models

[testExportExampleModelBasicEObjectsToXlsxNoExportMetadataNorAddMappingTable2792456214029170084.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775280/testExportExampleModelBasicEObjectsToXlsxNoExportMetadataNorAddMappingTable2792456214029170084.xlsx)
[testExportExampleModelBasicEObjectsToXlsxNonContainmentDisabled7887152567753493222.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775281/testExportExampleModelBasicEObjectsToXlsxNonContainmentDisabled7887152567753493222.xlsx)
[testExportExampleModelBasicResourceToXlsxNonContainmenEnabled10318917623464411669.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775282/testExportExampleModelBasicResourceToXlsxNonContainmenEnabled10318917623464411669.xlsx)
[testExportUtilModelEObjectsToXlsxNonContainmentDisabled9620885381954038949.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775283/testExportUtilModelEObjectsToXlsxNonContainmentDisabled9620885381954038949.xlsx)
[testExportTreesModelEObjectsToXLSXNonContainmenEnabled3499081249584906437.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775284/testExportTreesModelEObjectsToXLSXNonContainmenEnabled3499081249584906437.xlsx)
[testExportUtilModelEObjectsToXlsxNoExportMetadataNorAddMappingTable16035788605364877680.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775285/testExportUtilModelEObjectsToXlsxNoExportMetadataNorAddMappingTable16035788605364877680.xlsx)
[testExportUtilModelResourceToXlsxNonContainmenEnabled11436794266587490758.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775286/testExportUtilModelResourceToXlsxNonContainmenEnabled11436794266587490758.xlsx)
[testSaveXLSX2649416721126831944.xlsx](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775287/testSaveXLSX2649416721126831944.xlsx)


---
### ODS documents generated off 3 different models

[testExportUtilModelEObjectsToOdsNonContainmentEnabled4911314555720570965.ods](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775291/testExportUtilModelEObjectsToOdsNonContainmentEnabled4911314555720570965.ods)
[testSaveResourceToOds2014464433647793194.ods](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775292/testSaveResourceToOds2014464433647793194.ods)

[testExportExampleModelBasicResourceToOdsExportNonContainmentDisabled18261744029070465176.ods](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775288/testExportExampleModelBasicResourceToOdsExportNonContainmentDisabled18261744029070465176.ods)
[testExportExampleModelBasicResourceToOdsExportNonContainmentEnabled4207044031986672779.ods](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775289/testExportExampleModelBasicResourceToOdsExportNonContainmentEnabled4207044031986672779.ods)
[testExportTreesModelEObjectsToODSNonContainmentEnabled5665848752922577889.ods](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/12775290/testExportTreesModelEObjectsToODSNonContainmentEnabled5665848752922577889.ods)

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/EMFExportOptions.java:
```diff
@@ -21,23 +21,16 @@ public interface EMFExportOptions {
-	// export non-containment references, in addition to containment references (exported by default)
+	// export non-containment references, in addition to containment references
+	// (exported by default)
 	String OPTION_EXPORT_NONCONTAINMENT = "EXPORT_NONCONTAINMENT";
```
In order to avoid accidentally Exporting a whole DB, the default for this should be false. 

If it is false, I further more would change the definition of NonContainment export a little bit:

If this is false, the following must apply:

```
Family simpsonFamily = createSimpsonFamily(basicFactory);
xmiResource.getContents().add(simpsonFamily);
exporterService.exportResourceTo(xmiResource, fileOutputStream, 
				Map.of(
						EMFExportOptions.OPTION_LOCALE, Locale.GERMANY						
					)
				);
```
Output: only the Family is exported, but no Persons referenced inside it.

```
Family simpsonFamily = createSimpsonFamily(basicFactory);
xmiResource.getContents().add(simpsonFamily);
xmiResource.getContents().add(simpsonFamily.getFather());
xmiResource.getContents().add(simpsonFamily.getMother());
xmiResource.getContents().addAll(simpsonFamily.getChildren());
exporterService.exportResourceTo(xmiResource, fileOutputStream, 
				Map.of(
						EMFExportOptions.OPTION_LOCALE, Locale.GERMANY						
					)
				);
```
Output: Family and The Persons will be exported, as they are part of the Resource to export.

ideas-into-software replied:
Non-containment is defined on EMF model itself, i.e.: 

```
<eStructuralFeatures xsi:type="ecore:EReference" (...) containment="true" (...)>
```
			
 .. which then allows to check if given EReference is a containement via: `org.eclipse.emf.ecore.EReference.isContainment()`
 		 		
That is how it was understood so far, exactly as per what EMF's definition of containment.

If we attempt to change such, shouldn't we simply name it in a different way, so at not cause confusion / clash? also, how would we check if we'd not be able to use EMF's built-in way (i.e. `org.eclipse.emf.ecore.EReference.isContainment()`) ?

juergen-albert replied:
I have the feeling you are mixing things up here.

the `father, `mother` and `children` references are Non-Containments meaning:

```
<eStructuralFeatures xsi:type="ecore:EReference" (...) containment="false" (...)>
```

Thus `org.eclipse.emf.ecore.EReference.isContainment()` will always return `false`.  

So, if the `EXPORT_NONCONTAINMENT` option is not set, you need to call `get(<feature>, false)`, which means don't resolve the Object. If the return Object is no Proxy it should only be exported if it is part of the Resources content list. If it is a Proxy, you need to look if the URI matches the Resources you are exporting. If this is the case, you can use the Object you would get from `Resource.getEObject(<URI Fragment))`.

ideas-into-software replied:
I summarized my understanding of how 'containment' is used / understood in EMF and how it's used in the EMF Exporters' code currently. 

I was referring to example you provided, i.e.: 

> 	(...) If this is false, the following must apply: (...)
> 
> 	Family simpsonFamily = createSimpsonFamily(basicFactory);
> 	xmiResource.getContents().add(simpsonFamily);
> 	xmiResource.getContents().add(simpsonFamily.getFather());
> 	xmiResource.getContents().add(simpsonFamily.getMother());
> 	xmiResource.getContents().addAll(simpsonFamily.getChildren());
> 	exporterService.exportResourceTo(xmiResource, fileOutputStream, 
> 					Map.of(
> 							EMFExportOptions.OPTION_LOCALE, Locale.GERMANY						
> 						)
> 					);
> 
> 	Output: Family and The Persons will be exported, as they are part of the Resource to export.
> 
> 	(...)


In the `org.gecko.emf.osgi.example.model.basic` model, `Family` EClass defines 3 references (`father`, `mother` and `children`) which are marked in that EMF model as non-contained, i.e. they exist separately from `Family`. 

However, the example you provided (quoted above) I understood as: even though those 3 references are non-containment, if we add them to Resource, they are still exported, even though `EXPORT_NONCONTAINMENT` option is set to false (vide (...) _If this is false, the following must apply:_ (...)). 

Perhaps `Resource` has a different understanding of what containement means, and you're referring to that implicitly - I am referring to how that is understood in EMF model and code written so far. 

I need to know exactly what you mean.

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/EMFExportOptions.java:
```diff
@@ -21,23 +21,16 @@ public interface EMFExportOptions {
-//	String OPTION_FREEZE_HEADER_ROW = "FREEZE_HEADER_ROW";
 
+	// show URIs instead of IDs
+	String OPTION_SHOW_URIS = "SHOW_URIS";
```
I was wrong on my view on what the URIs would look like, as I had been blinded by the way EMFJson constructs its refrences (see `org.eclipse.emfcloud.jackson.databind.ser.EcoreReferenceSerializer.getHRef(...)`). Thus using URIs instead fo IDs does not make sense with the way it is done now. It might work for Objects with an ID, but for Objects without an ID feature, the resulting fragment does not look like I expected and thus the whole thing can be reverted.

We can use it to control if any ref fields should be written, that point to a non containment reference outside our exported list.

Example:
```
SHOW_URIS == true
//result would be

family.id, name,  family.father._ref, family.mother._ref,...
```

```
SHOW_URIS == false
//result would be

family.id, name
```

ideas-into-software replied:
I added a separate option and called it `SHOW_URIS` because URIs are only displayed in `ref_` fields (if that option is turned on).
 
Underneath, IDs are still being used, because I would not be able to automate this properly / non-deterministic behavior (different formats, lack of URI / incomplete URI / lack of unique identifier / etc) would make it very difficult to work. That is why I did not agree with you and continued using IDs underneath so all other features can work properly. It is only at the last stage where data collected in matrices is output to specific format (CSV, XLSX, ODS, etc.) that URIs can be displayed instead of IDs and only in `ref_` fields, otherwise IDs are displayed.
 
Having hopefully clarified this enough, do you still wish to have this removed ? If you wish to have option to control whether `ref_` fields are shown, shouldn't we name this in a different way - e.g. `SHOW_REFS`? Logic for this would be much different than that for `SHOW_URIS` / there would be no reuse as it's at a different stage that this would need to be handled - i.e. both at the time of creating headers, then data as certain columns (`ref_`) would need to be skipped.

juergen-albert replied:
Given what I said in my comment [here](https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370485601): if the `OPTION_EXPORT_NONCONTAINMENT` is not true, we must only export Objects in our given List. This means that every Object outside of our resulting  exported document would become a ref field, as EMF usually does for non Containment References in that exact case. 

This Option should control if such References should be exported or not. `SHOW_REFS` as a replacement name for that would be a good idea, as we call them refs.

ideas-into-software replied:
Exapnding on my understanding of containment / non-containment, this only denotes whether given `EClass` is contained / not-contained by class referencing it (which becomes a container for such EReferences); so e.g. when serializing this to MongoDB, only one collection would be created, instead of separate collections for each non-contained referenced EClass.

`EObject`s passed to `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` can have both contained and non-contained EReferences - e.g. `org.gecko.emf.osgi.example.model.basic.Person` (see `org.gecko.emf.csv.tests.EMFCSVExporterTest.testExportExampleModelBasicResourceToCsvFlatModeDifferentRootObjectsNonContainmentDisabled(ServiceAware<EMFExporter>, BasicFactory, BasicPackage)` test case), which has both contained (`Contact`, `Tag`) and non-contained (`Address`, `Person`) references.

So, if I pass list of `org.gecko.emf.osgi.example.model.basic.Person` to `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` and specify `EXPORT_NONCONTAINMENT` as false, wouldn't that still control only if non-contained (`Address`, `Person`) references are exported ? Because that is exactly as per EMF's definition of such. 

Would `SHOW_REFS` in turn control in such case whether any references (contained or non-contained) are extracted and exported ? So, if I pass list of `org.gecko.emf.osgi.example.model.basic.Person` to `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` and specify `EXPORT_NONCONTAINMENT` as false and `SHOW_REFS` as false, the resulting export would export only `EAttribute` structural features defined on `org.gecko.emf.osgi.example.model.basic.Person` ? (no `EReference` structural features).

Alternatively, since there are two methods defined on `org.gecko.emf.exporter.EMFExporter` - the other is `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` - perhaps you wish to change implementation of that method only, i.e. such cases should only be handled if dealing with `Resource`, not with `EObject` list (`org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)`) ?

juergen-albert replied:
Yeah, the whole containment/noncontainment thing in EMF is a bit foreign to the normal Java model. Lets try to explain and answer your questions. This also takes your [comment ](https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370568408) into account.

First in general: For EMF a valid Object at the time of serialization must be contained somewhere. And where ever this place is, the root of such a containment hierarchy must be a Resource. So, in order to understand Noncontainment we would need to understand Containment.

Lets assume, that the Father Reference has containment set to true. and we have the following pseudo code:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
toSave.getContents().add(homer);

toSave,save();

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
  },
 { 
    type: "Family"
    name : "Adams"
  },
  { 
    type: "Person"
    name : "Homer"
  }
]
```

Thus, there can be only one "owner" of Homer. The moment you add homer to a Containment Reference, EMF will remove it from where ever he was owned before. The Resources Content List is a so called `ContainmentEList`, which means anything you add there will be removed on any other containment.

Now we assume again, that Father is non Containment and repeat the whole thing:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
toSave.getContents().add(homer);

toSave,save();

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
   father.ref : "Homer"
  },
 { 
    type: "Family"
    name : "Adams"
    father.ref : "Homer"
  },
  { 
    type: "Person"
    name : "Homer"
  }
]
```

Now both Families point to Homer in the same Document.

Other Version:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Resource homerResource = new Resource("file:/homer.bar");

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
homerResource .getContents().add(homer);

toSave,save();

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
   father.ref : "file:/homer.bar#Homer"
  },
 { 
    type: "Family"
    name : "Adams"
    father.ref : "file:/homer.bar#Homer"
  },
]
```

Now it points to the Homer Resource outside of the Document (may not be saved yet).



# Now to your specific Questions

> So, if I pass list of org.gecko.emf.osgi.example.model.basic.Person to org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>) and specify EXPORT_NONCONTAINMENT as false, wouldn't that still control only if non-contained (Address, Person) references are exported ? Because that is exactly as per EMF's definition of such.

That is correct. Only the Person with its contained Contacts and and Tags would be exported.

> Would SHOW_REFS in turn control in such case whether any references (contained or non-contained) are extracted and exported ? So, if I pass list of org.gecko.emf.osgi.example.model.basic.Person to org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>) and specify EXPORT_NONCONTAINMENT as false and SHOW_REFS as false, the resulting export would export only EAttribute structural features defined on org.gecko.emf.osgi.example.model.basic.Person ? (no EReference structural features).

No. Take my example from above as a baseline and add a mother as a containment feature:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Resource homerResource = new Resource("file:/homer.bar");

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

Person marge = new Marge();
simpsons.setMother(marge);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
homerResource .getContents().add(homer);

toSave,save("SHOW_REFS ",  false);

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
    mother: { 
      type : "Person",  
      name: "marge"
   }
 },
 { 
    type: "Family"
    name : "Adams"
  },
]
```

and also: 

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Person homer = new Homer();

Resource gomezResource = new Resource("file:/gomez.bar");

Person gomez= new Gomez();

simspons.setFather(homer);
adams.setFather(gomez);

Person marge = new Marge();
simpsons.setMother(marge);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
toSave.getContents().add(homer);

toSave,save("SHOW_REFS ",  false);

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
    father.ref : "Homer"
    mother: { 
      type : "Person",  
      name: "marge"
   }
 },
 { 
    type: "Family"
    name : "Adams"
  },
  { 
    type: "Person"
    name : "Homer"
  }
]
```

Thus "SHOW_REFS" false mean: Don't show Reference URIs to Objects that are not contained in any of the given Object or the Resource itself (see the gomze, that is not Contained in the same Resource.

ideas-into-software replied:
@juergen-albert

I realize you've worked with EMF for much longer, but EMF is quite simple, at least in the sense of how containment / non-containment is defined on the EMF model itself (`<eStructuralFeatures xsi:type="ecore:EReference" (...) containment="true" (...)>`), which can then be checked via `org.eclipse.emf.ecore.EReference.isContainment()`. 

You referred several times to `Resource` and, as I mentioned in previous comment (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370568408), "_Perhaps Resource has a different understanding of what containement means, and you're referring to that implicitly_".

There are two public methods defined on `org.gecko.emf.exporter.EMFExporter` (as reminded in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370569897): 

 * `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` and 
 
 * `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)`
 
When I pass a list of `EObject`s to `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)`, where from would I be able to determine all these additional criterions which you mention? There, I only have access to `org.eclipse.emf.ecore.EReference.isContainment()` (derived from whatever is defined in the EMF model itself) and `exportOptions` (in this particular case, options `EXPORT_NONCONTAINMENT` and `SHOW_REFS`).

As mentioned in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370569897, perhaps you wish to change implementation of `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` method only, i.e. such cases should only be handled if dealing with `Resource` (`org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` method), not with `EObject` list (`org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` method) ? If I can programmatically access / derive these criterions from `Resource` passed to `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)`, to properly automate handling such cases, that is.

juergen-albert replied:
> Perhaps Resource has a different understanding of what containement means, and you're referring to that implicitly

Yes, that is kind of correct. EObjects must be contained somewhere. Either directly in a resource or in an EObject with a proper containment reference, where the root is contained in a Resource. The basic check is: `EObject.eResource()` must return something. 

Looking at the the two methods you have I can say the following:

`org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` - here it is fairly easy as you would only need to check if any Object you may encounter points to the given resource.

`org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` - here it is a bit harder. If you encounter a non containment Reference, you would need to get the Object it points to, without triggering a proxy resolve and then you would need to check if the EObject you have received is in the initial List you have gotten to export. If it is a Proxy, you would need to compare its URI, with the URIs of all the given Objects in your List.

The `exportResourceTo`, is the one I and the Customer really wants and needs. 
I see the other one as useful for testing, but I have the feeling, that it otherwise creates more complexity.

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1766 @@
+
+	protected static final String ECORE_PACKAGE_NAME = "ecore";
+
+	protected static final String ID_COLUMN_NAME = "Id";
```
After giving the whole id/pseudo id topic some more thought, I'd make the following request:
We stick to your mechanism in using the ids or pseudoIds if no ID is available. We should however add a `_id` header for all Objects, regardless if they have and ID field nor not. This would be in line with what mongoEMF, EMFJson and a few others are doing.

ideas-into-software replied:
As you know, currently, if given `EClass` does not have an ID, this "pseudo ID" mechanism kicks in and such "pseudo ID' is generated. In matrix for that given `EClass`, additional column is created, named `id`, and there those "pseudo IDs" are output. Then, during output of data collected in matrices to specific format (CSV, XLSX, ODS, etc.), that `id` column is transferred. So, this is already done, but named differently (i.e. `id`, not `_id` like you mentioned).
 
In addition, metadata sheet for such `EClass` contains additional entry for such "pseudo ID" field, i.e. "No default ID present, therefore pseudo-ID was generated for identification purposes". See screenshots attached.
 
If I rename this to `_id` (from `id`, like currently), would that satisfy your requirement ? 

![Screenshot from 2023-10-24 17-45-36](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/5586faa0-eab0-48c3-bb4c-3ba2b63cad3d)

![Screenshot from 2023-10-24 17-45-44](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/8c008dd5-d5a2-4793-aea8-89b4579da168)

juergen-albert replied:
I know and I do understand that. What I'm getting to is the following:

We should always and for every Object have an _id field. This is populated either by the value of the ID Feature or with your pseudo id.  For the simplified family example this would look as follows:

```
family._id, family.id, name
id1,id1,Simpsons
```

With this, there would be always a defined and known id field that can be used, regardless of the name of the id feature. MongoEMF has the same mechanism and this works great.

ideas-into-software replied:
What I meant is that there already ALWAYS is such field and it's named `id`. It is populated with value of `id` attribute of given EObject, if such is defined, otherwise with pseudo-id. But it's already there and it's named `id`. 

That is why I asked whether I should rename `id` to `_id`, since that functionality is already there. From your answer, I understand I should add ANOTHER column, named `_id`, which will essentially be exactly the same as `id` / contain same exact data. Is that correct?

juergen-albert replied:
In EMF any Attribute can be an id field. We name them id by convention. Thus if we always have and _id field in the export it would a be clearer that this is a generated field. If we use the family example again and remove the id attribute and declare the name as ID field the whole thing would look as follows:

```
family._id, family.name
Simpsons, Simpsons
```

Might seem as overhead, but I think this is clearer.

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1766 @@
+
+	protected static final String ID_ATTRIBUTE_NAME = "id";
+
+	protected static final String REF_COLUMN_PREFIX = "ref_";
```
Please change this to a sufix. A Ref should become e.g.`family.father._ref`. This would be closer to e.g. mongoEMF (`family.father._eProxyUri`) and EMFJson (`family.father.$ref`). With this we also establish the pattern, that _ marks a reserved property, that is not part of the actual model.

ideas-into-software replied:
OK, no problem.

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1766 @@
+
+	// maps EObjects' IDs to names of matrices, so those can be looked up e.g. when
+	// constructing links
+	protected final Map<String, String> eObjectIDToMatrixNameMap;
```
We must make the services, that result from this abstract class, stateless and hand the lists and maps used for caching everywhere we go (maybe via a Context, to not explode the method signatures). 

Even though the services I saw are prototyped, this is a Problem. For one, the user of the Services does not know that the service is stateful and thus, does not necessary use a new instance for eache usage. This creates memory leaks and also may produce concurrency issues.

This already happens e.g. in the `EMFXLSXResourceFactoryConfigurator`, which uses one instance for as long as it exists.

ideas-into-software replied:
All these data structures are reset via `org.gecko.emf.exporter.AbstractEMFExporter.resetState()` with each call to public methods.

As you know, these data structures are needed to speed up (hundreds of times speed up) processing, to be able to construct lookup indexes / references / links / etc.
I moved them to instance level as there are already 6 such data structures in `org.gecko.emf.exporter.AbstractEMFExporter` and even more in format-specific exporters (i.e. additional 4 in `org.gecko.emf.exporter.csv.impl.EMFCSVExporter`, etc.).
 
Is resetting state on each call to public method (via `org.gecko.emf.exporter.AbstractEMFExporter.resetState()`) an acceptable solution for you or should I proceed with refactor and introduce a parameter object where these data structures are held ?

juergen-albert replied:
Not sure if this would cut it. I would Expect that the Exporter can handle concurrent calls to the export method on the same instance. I believe that even a resetState method will not make it Thread safe.

ideas-into-software replied:
I will introduce a parameter object then. 

I was under the impression, that since a lot of EMF and Gecko stuff is not thread safe, and here I'm using scope `PROTOTYPE`, this will be sufficient. But I will change this as you prefer then.

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS)  (c.d.)

 - Make 'EXPORT_NONCONTAINMENT' option, as well as dependent options, 
false by default
 
 - Add 'show REFs' option, and skip 'ref_' columns if turned off
 
 - Rename 'id' column to '_id'
 
 - Change 'ref_' from prefix to suffix
 
 - Thread safety improvements

 - Various other fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
@juergen-albert

With this most recent commit (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23/commits/c8bb0b59df34285671f0a585467cbf8d0c3ad061) following were taken care of: 

 - Make `EXPORT_NONCONTAINMENT` option, as well as dependent options, false by default
 
 - Add `SHOW_REFS` option, and skip `ref_` columns if turned off
 
 - Rename `id` column to `_id`
 
 - Change `ref_` from prefix to suffix
 
 - Thread safety improvements


Attached please find output from all 3 exporters, with both 'show REFs' turned off and turned on.:
• CSV (flat mode) documents generated off 3 different models
• CSV (ZIP mode) documents generated off 3 different models
• XLSX documents generated off 3 different models
• ODS documents generated off 3 different models


Please review those and in the meantime I will be taking care of one last item from your earlier requests, i.e. pertaining to `Resource`.

[testExportToCsvShowREFsOFF.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13181100/testExportToCsvShowREFsOFF.zip)
[testExportToCsvShowREFsON.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13181101/testExportToCsvShowREFsON.zip)

[testExportToXlsxShowREFsOFF.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13181114/testExportToXlsxShowREFsOFF.zip)
[testExportToXlsxShowREFsON.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13181116/testExportToXlsxShowREFsON.zip)

[testExportToOdsShowREFsOFF.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13181129/testExportToOdsShowREFsOFF.zip)
[testExportToOdsShowREFsON.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13181131/testExportToOdsShowREFsON.zip)

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS) (c.d.)

 - Preprocess contents of 'Resource' when export non-containment option
is disabled

 - Various other fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
@juergen-albert

With this most recent commit (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23/commits/416dc9a0fdd62718c04f657637f663ce7d4df736), preprocessing contents of `Resource` when export non-containment option is disabled was taken care of; also please refer to earlier comment pertaining to commit from 3 days ago (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#issuecomment-1781593148) - there, all the other additional requirements / changes were taken care of.

Attached please find output from all 3 exporters:
• CSV (flat mode) documents generated off 3 different models
• CSV (ZIP mode) documents generated off 3 different models
• XLSX documents generated off 3 different models
• ODS documents generated off 3 different models


For examples of output with 'show REFs' option turned off, please refer to earlier comment pertaining to commit from 3 days ago (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#issuecomment-1781593148)


**Please review and clarify / verify the following**:

**A**) **RE**: "_you need to look if the URI matches the Resources you are exporting_" (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370485601)

 Clarify which part of URI should match - entire URI, last segment, fragment ? Current implementation @ `org.gecko.emf.exporter.AbstractEMFExporter.uriMatches(URI, URI, boolean)`


**B**) **RE**: "_So, if the EXPORT_NONCONTAINMENT option is not set, you need to call get(<feature>, false), which means don't resolve the Object. If the return Object is no Proxy it should only be exported if it is part of the Resources content list. If it is a Proxy, you need to look if the URI matches the Resources you are exporting._" (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370485601) **and** "_If you encounter a non containment Reference, you would need to get the Object it points to, without triggering a proxy resolve and then you would need to check if the EObject you have received is in the initial List you have gotten to export. If it is a Proxy, you would need to compare its URI, with the URIs of all the given Objects in your List._" (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1371777706)

 Verify my understanding of this requirement - current implementation @ `org.gecko.emf.exporter.AbstractEMFExporter.preProcessResourceEObject(EObject, EReference, Set<String>, PreProcessedResourceEObjectsDTO, Map<Object, Object>)`; additional test cases @ `org.gecko.emf.xlsx.tests.EMFXLSXExporterTest`

[testExportToXlsx_20231029.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13198242/testExportToXlsx_20231029.zip)

[testExportToCsv_20231029.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13198243/testExportToCsv_20231029.zip)

[testExportToOds_20231029.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13198245/testExportToOds_20231029.zip)

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1957 @@
+
+		// TODO: vary processing depending on whether 'exactMatch' is set/true
+
+		return (resourceEObjectURI.toString()).equalsIgnoreCase(eObjectURI.toString());
```
> Clarify which part of URI should match - entire URI, last segment, fragment ? Current implementation @ org.gecko.emf.exporter.AbstractEMFExporter.uriMatches(URI, URI, boolean)

Usually a Resource URI has no fragment and no Query. As one can set any URI to a Resource thouge, it becomes a bit more complex. Thus I would suggest the following logic:


```java
//given what we discussed before, all theObjects to export should have the same resource and thus only one URI what be the candidate to compare against.
URI eObjectUri; 
if(!resourceURI.hasQuery()){
    eObjectUri = eObjectUri.trimQuery();
}

if(!resourceURI.hasFragment()){
    eObjectUri = eObjectUri.trimFragment();
}
return resourceURI.equals(eObjectUri);
```

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1957 @@
+
+			} else if (eReference.isMany()) {
+
+				if (((List<EObject>) rawEObject).isEmpty()) {
```
This if is unnecessary. In EMF you will always get an empty list for a many reference that is not populated. Thus the remove will not do anything and you can always assume that you will get a non null list.

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1957 @@
+				List<EObject> processableEObjects = new ArrayList<>();
+				List<EObject> unProcessableEObjects = new ArrayList<>();
+
+				for (EObject maybeProcessableEObject : (List<EObject>) rawEObject) {
```
Please use the following way to iterate over this list:

```java
for (int i = 0; i < list.size(); i++) {
    EObject eObject = list.basicGet(i);
    //This will prevent any proxy to be resolved. Usually the list will be an EcoreEList, where the get(i) causes the proxy to beeing resolved (aka loaded e.g. from the database)
}

```

juergen-albert commented on org.gecko.emf.exporter/src/org/gecko/emf/exporter/AbstractEMFExporter.java:
```diff
@@ -0,0 +1,1957 @@
+
+		preProcessedEObjectsDTO.rawResourceEObjects.addAll(resource.getContents());
+
+		preProcessedEObjectsDTO.resourceEObjectsURIs
```
You should store the resource as well, because then you can simply compare `eResource()` of an EObject (if present) with the given resource without doing complex URI matching.

juergen-albert commented:
Another thing:

Looking at testExportUtilModelEObjectsToCsvFlatModeNonContainmenEnabled114535766836992159.csv I see the following headers: `Request_id,from,to,page,...`

I would expect to se the following: `_id,id,from,to,page,...`

Having and `_id` must not remove the actual Attribute marked as id. Additionally the case with the request is intersecting, as the attribute with the name `id`, is not marked as the ID value. As the Model defines the truth, it can't be assumed that it is the actual ID.

BTW: I realized that I made a mistake in my example on https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370517271. In that example I marked uselessly added the `family.` as a prefix to the `id` and `_id`, which is unnecessary for a root object.

juergen-albert commented:
testExportExampleModelBasicResourceToCsvFlatModeDifferentRootObjectsNonContainmentDisabled16487394186529543123.csv

The result looks largely correct. Besides the id issue I pointed out [here](https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#issuecomment-1784922966), the proxy URIs are now shown correctly. The headers though are a bit inconsistent with non containment lists though: `relatives._ref.0` must be `relatives.0._ref`

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS) (c.d.)

 - Improvements for preprocessing contents of 'Resource' - URI matching,
iterating over list of EOBjects without triggering resolve, and other

 - '_id' column as an additional column, not replacement for 'id' column

 - Position of '_ref' suffix at the end for many references' column
header names

 - Various other fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
@juergen-albert

With this most recent commit (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23/commits/c93ba651a3077fd2a80f76d77c3a0e46f9639801), all of the changes you requested on 30.10.2023 are taken care of.

Attached please find output from all 3 exporters:
• CSV (flat mode) documents generated off 3 different models
• CSV (ZIP mode) documents generated off 3 different models
• XLSX documents generated off 3 different models
• ODS documents generated off 3 different models


Please review and confirm this is ready to go, and I'll proceed with integrating this into MDO project, as previously discussed.

[testExportToXlsx_20231101.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13222042/testExportToXlsx_20231101.zip)
[testExportToCsv_20231101.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13222043/testExportToCsv_20231101.zip)
[testExportToOds_20231101.zip](https://github.com/geckoprojects-org/org.gecko.emf.utils/files/13222044/testExportToOds_20231101.zip)

juergen-albert commented:
LGTM. Can you update the PR, so the merge Conflicts are resolved? With this I would merge and release it.

Commit:
Merge remote-tracking branch 'upstream/jakarta' into emf-xlsx-ods-csv-exporters

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>
